stone/target/python_rsrc/stone_serializers.py

```diff
@@ -24,11 +24,471 @@
 except (SystemError, ValueError):
     # Catch errors raised when importing a relative module when not in a package.
     # This makes testing this file directly (outside of a package) easier.
-    import stone_validators as bv  # type: ignore
+    import stone_validators as bv # type: ignore
+
+#-------------------------------------------------------------------------
+class StoneEncoderInterface(object):
+    """
+    Interface defining a stone object encoder.
+    """
+
+    def encode(self, validator, value):
+        """
+        Validate ``value`` using ``validator and return the encoding.
+
+        Args:
+            validator: the stone validator used to validate ``value``
+            value: the object to encode
+
+        Returns:
+            This is implementation-defined. The convention is to return an
+            encoded object where ``out_file`` is ``None``, and return
+            ``None`` if ``out_file`` is provided (and supported)
+
+        Raises:
+            stone_validators.ValidationError: Raised if ``value`` (or one
+                of its sub-values).
+        """
+        raise NotImplementedError
+
+    def encode_to_file(self, validator, value, out_file):
```

Comment:
Okay, I'm sold. The current implementation uses `json.dumps` (as opposed to `json.dump(…, io.BytesIO())`), and `json.dumps` is almost 10× faster (at least according to my tests¹):

```sh
% python …/test.py
running <function mk_f_dump at 0x10f6e9c08> (run 1 of 10)...1.339 s
running <function mk_f_dump at 0x10f6e9c08> (run 2 of 10)...1.478 s
running <function mk_f_dump at 0x10f6e9c08> (run 3 of 10)...1.353 s
running <function mk_f_dump at 0x10f6e9c08> (run 4 of 10)...1.339 s
running <function mk_f_dump at 0x10f6e9c08> (run 5 of 10)...1.334 s
running <function mk_f_dump at 0x10f6e9c08> (run 6 of 10)...1.327 s
running <function mk_f_dump at 0x10f6e9c08> (run 7 of 10)...1.319 s
running <function mk_f_dump at 0x10f6e9c08> (run 8 of 10)...1.308 s
running <function mk_f_dump at 0x10f6e9c08> (run 9 of 10)...1.319 s
running <function mk_f_dump at 0x10f6e9c08> (run 10 of 10)...1.317 s
<function mk_f_dump at 0x10f6e9c08>: 1.343 avg s (10 runs)
running <function mk_f_dumps at 0x10f6e9c80> (run 1 of 10)...0.167 s
running <function mk_f_dumps at 0x10f6e9c80> (run 2 of 10)...0.162 s
running <function mk_f_dumps at 0x10f6e9c80> (run 3 of 10)...0.168 s
running <function mk_f_dumps at 0x10f6e9c80> (run 4 of 10)...0.158 s
running <function mk_f_dumps at 0x10f6e9c80> (run 5 of 10)...0.162 s
running <function mk_f_dumps at 0x10f6e9c80> (run 6 of 10)...0.170 s
running <function mk_f_dumps at 0x10f6e9c80> (run 7 of 10)...0.160 s
running <function mk_f_dumps at 0x10f6e9c80> (run 8 of 10)...0.174 s
running <function mk_f_dumps at 0x10f6e9c80> (run 9 of 10)...0.161 s
running <function mk_f_dumps at 0x10f6e9c80> (run 10 of 10)...0.187 s
<function mk_f_dumps at 0x10f6e9c80>: 0.167 avg s (10 runs)
```

Removed `encode_to_file` and all orphaned `out_file=None` arguments.

---

¹ `test.py` source:

```py
# test.py
from __future__ import absolute_import, division, print_function, unicode_literals

import functools
import io
import json
import time

def mknode(depth):
    d = {}
    if depth > 0:
        for i in range(depth, depth + 10):
            if i & 0x1:
                i = str(i)
            d[i] = mknode(depth - 1)
        for i in range(depth + 10, depth + 20):
            if not i & 0x1:
                i = str(i)
            d[i] = [ mknode(depth - 1) for _ in range(depth) ]
    for i in range(depth + 20, depth + 40):
        if i & 0x1:
            v = str(i)
        else:
            v = i
        if i & 0x2:
            k = str(i)
        else:
            k = i
        d[k] = v
    return d

def time_f(mk_f, n=10, d=3):
    tot = 0
    for i in range(n):
        candidate = mknode(d)
        f = mk_f(candidate)
        print('running {} (run {} of {})...'.format(mk_f, i + 1, n), end='')
        s = time.time()
        f()
        e = time.time()
        t = e - s
        print('{:.3f} s'.format(t))
        tot += t
    print('{}: {:.3f} avg s ({} runs)'.format(mk_f, tot / n, n))

def mk_f_dump(obj):
    fp = io.BytesIO()
    return functools.partial(json.dump, obj, fp)

def mk_f_dumps(obj):
    return functools.partial(json.dumps, obj)

for mk_f in mk_f_dump, mk_f_dumps:
    time_f(mk_f)
```